src/coreclr/tools/Common/TypeSystem/IL/NativeAotILProvider.cs

```diff
@@ -14,6 +16,13 @@ namespace Internal.IL
 {
     public sealed class NativeAotILProvider : ILProvider
     {
+        private readonly CompilationModuleGroup _compilationGroup;
+
+        public NativeAotILProvider(CompilationModuleGroup group)
```

Comment:
Could you instead `#ifdef READYTORUN` the InterlockedIntrinsics to not take CompilationModuleGroup outside READYTORUN?

ILProvider taking a dependency on ILCompiler stuff is odd and confusing. Nothing under Common\TypeSystem should reference ILCompiler namespace as a test for layering violations.